### PR TITLE
strict_dns support for cf apps from registry

### PR DESCRIPTION
**Pull request by GregHanson**
finds instances from registry and build a static dns cluster for them in envoy.json file

### CODE REVIEW

**Comment by rshriram** [comment]
Hold off on this please. It requires changes at Envoy level as well. This alone won't suffice

**Comment by rshriram** [comment]
You might want to continue this again. Basically if we have CDS, and we return strict dns, this should work with the auto_host_rewrite option set in the route. See Envoy pr

File: sidecar/proxy/envoy/manager.go [hunk 1]
```diff
@@ -330,65 +333,67 @@ func BuildClusters(instances []*api.ServiceInstance, rules []api.Rule, tlsConfig
 	rules = AddDefaultRouteRules(rules, instances)
 
 	backends := make(map[string]*api.Backend)
+	staticClusters := make(map[string]*api.Backend)
+	instanceByClusterMap := make(map[string][]*api.ServiceInstance)
```
Comment (ijsnellf) [inline sidecar/proxy/envoy/manager.go]: `instancesByClusterMap`?

File: sidecar/proxy/envoy/manager.go [hunk 2]
```diff
@@ -330,65 +333,67 @@ func BuildClusters(instances []*api.ServiceInstance, rules []api.Rule, tlsConfig
 	rules = AddDefaultRouteRules(rules, instances)
 
 	backends := make(map[string]*api.Backend)
+	staticClusters := make(map[string]*api.Backend)
```
Comment (ijsnellf) [inline sidecar/proxy/envoy/manager.go]: The logic from line 336 to 409 could be simplified by making `staticClusters` a set.

File: sidecar/proxy/envoy/manager.go [hunk 3]
```diff
@@ -403,15 +408,84 @@ func BuildWeightKey(service string, tags []string) string {
...
+
+	return cluster
+}
+
 // BuildRoutes builds routes based on rules.  Assumes at least one route for each service
 func BuildRoutes(ruleList []api.Rule, instances []*api.ServiceInstance) []Route {
 	SanitizeRules(ruleList)
 	rules := AddDefaultRouteRules(ruleList, instances)
 
+	staticClusters := make(map[string]*api.Backend)
```
Comment (ijsnellf) [inline sidecar/proxy/envoy/manager.go]: We never use the map value `*api.Backend`, so `staticClusters` would probably be more appropriate as a set: `map[string]struct{}`

**Review by ijsnellf — CHANGES_REQUESTED** [review]
Could be simplified a bit.

File: sidecar/proxy/envoy/manager.go [hunk 4]
```diff
@@ -330,65 +333,67 @@ func BuildClusters(instances []*api.ServiceInstance, rules []api.Rule, tlsConfig
...
-			}
+		clusters = append(clusters, cluster)
+	}
 
-			// Envoy outlier detection settings that complete circuit breaker
-			if backend.Resilience.SleepWindow > 0 {
-				cluster.OutlierDetection.BaseEjectionTimeMS = int(backend.Resilience.SleepWindow * 1000)
+	for clusterName, backend := range staticClusters {
+		cluster := buildCluster(clusterName, backend, tlsConfig)
+		cluster.Type = "strict_dns"
```
Comment (ijsnellf) [inline sidecar/proxy/envoy/manager.go]: `buildStaticCluster` function?

**Review by ijsnellf — APPROVED** [review]

**Pull request merged**